crypto/objects/obj_dat.c

```diff
@@ -221,112 +228,22 @@ void ossl_obj_cleanup_int(void)
     objs_free_locks();
 }
 
-/*
- * Requires that the ossl_obj_lock be held
- * if TSAN_REQUIRES_LOCKING defined
- */
-static int obj_new_nid_unlocked(int num)
+int OBJ_new_nid(int num)
 {
     static TSAN_QUALIFIER int new_nid = NUM_NID;
 #ifdef TSAN_REQUIRES_LOCKING
     int i;
 
+    ossl_obj_write_lock(1);
     i = new_nid;
     new_nid += num;
```

Comment:
Is there a reason this doesn't use atomics?

Reply:
So it seems that tsan_add() does things atomic, and when that's not available you need to take a lock.

Why doesn't tsan_add() take a lock? It seems we want the CRYPTO_atomic_add() functionality, but I guess we prefer a version that can be inlined. I think that tsan just sounds like a bad name for it. The tsan code was written for when the exact count wasn't that important, we don't mind that some values are lost. But in this case we want all values to be unique.

Reply:
I agree tsan is kind of lousy naming, but that seems a bit out of scope for this effort.